bin/android-functions.sh

```diff
@@ -4,6 +4,268 @@ declare root="$(cd "$(dirname "$(dirname "${BASH_SOURCE[0]}")")" && pwd)"
+  local _gh
+
+  for android_home_test in "${ANDROID_HOME_SEARCH_PATHS[@]}"; do
+    for sdk_man_test in "${ANDROID_SDK_MANAGER_SEARCH_PATHS[@]}"; do
+      if [[ -z "$_ah" ]]; then
+        write_log "v" "Checking $android_home_test/$sdk_man_test/sdkmanager$bat"
+        if [[ -f "$android_home_test/$sdk_man_test/sdkmanager$bat" ]]; then
+          _ah="$android_home_test"
+          _sdk="$sdk_man_test/sdkmanager$bat"
+          break
+        fi
+      fi
+    done
+  done
+
+  temp=$(mktemp)
+  for java_home_test in "${JAVA_HOME_SEARCH_PATHS[@]}"; do
+    # Try initial search in bin of known location before attempting `find`
+    write_log "v" "Checking $java_home_test/bin/javac$exe"
+    if [[ -f "$java_home_test/bin/javac$exe" ]]; then
+
+      test_javac_version "$java_home_test/bin/javac$exe" "$JDK_VERSION" ; r=$?
+      if [[ "$r" == "0" ]]; then
+        # subshell, output to file
+        echo "$(dirname "$(dirname "$javac")")" > "$temp"
+        echo "$java_home_test" > "$temp"
+        write_log "v" "Using a $javac"
+        break
+      else
+        write_log "v" "Ignoring a $java_home_test/bin/javac$exe $jc_v"
+        # configured JAVA_HOME is bad, unset to trigger new version install
+        [[ "$JAVA_HOME" == "$java_home_test" ]] && unset JAVA_HOME
+      fi
+    fi
+
+    write_log "v" "find $java_home_test -type f -name "javac$exe"' -print0 2>/dev/null | while IFS= read -r -d '' javac"
+    find "$java_home_test" -type f -name "javac$exe" -print0 2>/dev/null | while IFS= read -r -d '' javac
+    do
+      # break doesn't work here, check that we don't have a result
+      if [[ $(stat_size "$temp") == 0 ]]; then
+        write_log "v" "Found $javac"
+        test_javac_version "$javac" "$JDK_VERSION" ; r=$?
+        if [[ "$r" == "0" ]]; then
+          # subshell, output to file
+          write_log "v" "Using b $javac"
+          echo "$(dirname "$(dirname "$javac")")" > "$temp"
+        else
+          write_log "v" "Ignoring b $javac $jc_v"
+        fi
+      fi
+    done
+  done
+
+  if [[ $(stat_size "$temp") != 0 ]]; then
+    _jh=$(cat "$temp")
+    echo >&2 "jh: $_jh"
+  fi
+
+  echo -n > "$temp"
+
+  for gradle_test in "${GRADLE_SEARCH_PATHS[@]}"; do
+    write_log "v" "Checking $gradle_test/bin/gradle$bat"
+    if [[ -f "$gradle_test/bin/gradle$bat" ]]; then
+      echo "$gradle_test" > "$temp"
+      break
+    fi
+
+    write_log "v" "find $gradle_test -type f -name 'gradle' -print0 2>/dev/null | while IFS= read -r -d '' gradle"
+    find "$gradle_test" -type f -name "gradle$bat" -print0 2>/dev/null | while IFS= read -r -d '' gradle
+    do
+      # break doesn't work here, check that we don't have a result
+      if [[ $(stat_size "$temp") == 0 ]]; then
+        write_log "v" "Found $gradle"
+        # subshell, output to file
+        echo "$(dirname "$(dirname "$gradle")")" > "$temp"
+      fi
+    done
+  done
+
+  if [[ $(stat_size "$temp") != 0 ]]; then
+    _gh=$(cat "$temp")
+  fi
+  rm "$temp"
```

Comment:
should this prompt? (instead of `rm -f`)

Reply:
Can you elaborate?

Reply:
`rm` with `-f` prompts user if they are sure they want to remove the file:

```sh
$ temp=$(mktemp)
$ rm "$temp"
rm: remove regular empty file '/var/folders/gk/v3ww7l_x6xvgk0j466j505880000gn/T/tmp.1tPD7PWbui'?
```

Reply:
Which `rm` are you using? Mine doesn't operate that way
```bash
% temp=$(mktemp)                                                                                                                      
% rm $temp
% echo $?
0
```

Reply:
Hmm same one as always:
https://www.unix.com/man-page/posix/1p/rm/